### ISSUE TITLE
Better documentation on the meaning of "PATH"

### ISSUE THREAD

**Issue by kckrinke** [zecure/shadowd_php#26]
I've found the docs to be quite thin on defining what exactly the PATH things are, and how they're used specifically. For example, the `GET` key and usage for example is not mentioned anywhere.

Instinctively I'd want to configure things like:
```
[
    {
        "path": "GET|REQUEST_URI|^\/something/??[a-zA-Z0-9]{0,10}\/??$",
        "caller": "*",
        "min_length": 10,
        "max_length": 22,
        "filter": 7
    },
    ...
]
```

Where can I find more detailed information on how to manipulate/use the values?

(Just realized this is probably not the best place for the issue report. Let me know where to repost if this isn't a good venue.)

**Comment by zit-hb** [comment]
The `path` of a parameter is an identifier. In general all parts between the dashes can be arbitrary (but they are not). The first part specifies the source of the user input (e.g. in the case of PHP: `GET` for `$_GET`, `REQUEST` for `$_REQUEST`, `COOKIE` for `$_COOKIE`, ...). After that every part is a key and represents a new dimension in an array.

If you create rules there is only one special character, the wildcard: `*`. So if you have `GET|foo|*` this would match `?foo[]=x&foo[][z]=y`.

After using Shadow Daemon for some years I also think it would be nice to have the full power of regular expressions when trying to match a rule. Great care has to be taken to avoid creating a bottleneck though (e.g. ReDoS). It is unlikely that this will be added to Shadow Daemon 2. If I ever start Shadow Daemon 3 it would certainly support regular expressions though.

Good idea to add it to the documentation, I will do that in a bit.

**Comment by kckrinke** [comment]
So `path` and `resource` aren't the same thing. How is resource filtering handled?

**Comment by zit-hb** [comment]
Yes, they are not the same. `resource` for a request is only stored because it helps a human auditor in the web interface to classify it more easily. It is not used for the analysis at all.

`caller` is used to specify the "resource" accurately. By default it is the PHP file that is executed on a request. In general that makes most sense, unless you are using a framework like Symfony and routing everything through one file. In that case you can use the [connector config to set the caller to PHP_SELF](https://github.com/zecure/shadowd_php/blob/v2.0.2/misc/examples/connectors.ini#L55), that worked quite nicely for me so far.

I also recommend to have a look here to see it in action: https://demo.shadowd.zecure.org/blacklist/rules

**Comment by kckrinke** [comment]
For sites that rely almost entirely on the path portion of the REQUEST_URI field, how does one validate the input? Or perhaps I'm not getting how the REQUEST_URI is parsed for threats in some other way that makes this a non-issue?

Example:
- `/pages/category/urlencoded-thing-that-could-be-malicious`
- there are no actual query parameters used at all
- PHP_SELF is always `/index.php`

**Comment by webboty** [comment]
@zit-hb this is interesting and adds to the transparency to understand how shadowd works.

Just to confirm ... in order to have the "path/request" analysed correctly i need to modify the caller ... 

In my laravel app only request_uri contains a useful path ... everything else just points to index.php
So caller is set to request_uri ... correct?

See screenshot attached for the environment variables available. 

![2021-10-23_13-56-39](https://user-images.githubusercontent.com/47084177/138555257-2a039002-2a62-4f4c-82e4-7d0fb06a0143.png)

**Comment by zit-hb** [comment]
@webboty `REQUEST_URI` is not a good value for callers because it also contains the query string, i.e. `?company_id=1` in this case.

It is interesting that `PHP_SELF` does not contain the route like it does for Symfony (or at least did when I tested it the last time). For now I would recommend you to stick to the default value then. I will try to check out Laravel in the next days to see how I can solve that problem the best.
